CORDA-3596 Record flow metadata

lankydan opened the pull request:
Record flow metadata during the zero'th checkpoint that occurs before
calling the flow's `call` function.

This required adding an RPC call's arguments to the `InvocationContext`
that gets created. These arguments are then accessible within the
statemachine and from the `Checkpoint` class. The arguments are then
extracted when recording a flow's metadata inside of
`DBCheckpointStorage`.

Updated the size of the started by column to 128 since it was not long
enough to hold the fully qualified class of a service that started a
flow.

lankydan committed:
CORDA-3596 Record flow metadata

Record flow metadata during the zero'th checkpoint that occurs before
calling the flow's `call` function.

This required adding an RPC call's arguments to the `InvocationContext`
that gets created. These arguments are then accessible within the
statemachine and from the `Checkpoint` class. The arguments are then
extracted when recording a flow's metadata inside of
`DBCheckpointStorage`.

Updated the size of the started by column to 128 since it was not long
enough to hold the fully qualified class of a service that started a
flow.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -295,19 +288,25 @@ class DBCheckpointStorage(private val checkpointPerformanceRecorder: CheckpointP
-            invocationInstant = checkpoint.checkpointState.invocationContext.trace.invocationId.timestamp,
+            rpcUsername = context.principal().name,
+            invocationInstant = context.trace.invocationId.timestamp,
+            // [receivedInstant] makes no sense anymore as we have _started_ the flow
```
Seems like `invocationInstant` and `receivedInstant` are perhaps the same ?  Trying to think how they could differ...  maybe if `invocationInstant` came from the "other side" (e.g. from the RPC client, or other peer) then they could differ but not sure we transmit any such timestamp to use?

lankydan replied:
`invocationInstant` will come from the client. With this implementation, I would remove `receivedInstant` as it will always be the same as `startInstant`.

rick-r3 replied:
Yeah, that makes sense.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -295,19 +288,25 @@ class DBCheckpointStorage(private val checkpointPerformanceRecorder: CheckpointP
-            startType = DBCheckpointStorage.StartReason.RPC,
-            launchingCordapp = "this cordapp",
+            startType = context.getStartedType(),
+            initialParameters = context.getFlowParameters().storageSerialize().bytes,
```
Just checking we have everything we need if we were to try and implement a full restart "from first principals"?

i.e. we could ask the flow factory / starter to build a flow using `flowName` and `initialParameters`.

What format is the invocation context `arguments` list ?  Is it:

`arguments[0]` = `flowName: String`
`arguments[1]` = `params: List<Any?>`

?

lankydan replied:
Yes, that is the format. So we have everything we would need (flow name stored in another field).

lankydan replied:
Considering if I should pass around an `Array<Any?>` instead of `List<Any?>`. The arguments originally come in the form of an `Array` which I convert to a `List`. I suppose since we never need to alter the contents of the arguments, we could just stick to an `Array` (then I can remove the `toList()`s I have added).

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/rpc/RPCServer.kt:
```diff
@@ -372,10 +372,10 @@ class RPCServer(
                     when (arguments) {
                         is Try.Success -> {
+                            val context = artemisMessage.context(clientToServer.sessionId, arguments.value)
+                            context.invocation.pushToLoggingContext()
```
Does moving this here mean that in the failure branch below we haven't set some of the logging context?  Might be worth setting with empty args before the warning.

rick-r3 replied:
Any thoughts on this?

lankydan replied:
Just looking at it at the moment. The code suggests that you lose some of the logging info on the `Failure` code path. Was thinking a simple solution is to make `context` a `var` and set it with different arguments inside the `when` block.

lankydan replied:
```kotlin
val context: RpcAuthContext
when (arguments) {
    is Try.Success -> {
        context = artemisMessage.context(clientToServer.sessionId, arguments.value)
        context.invocation.pushToLoggingContext()
        log.debug { "Arguments: ${arguments.value.toTypedArray().contentDeepToString()}" }
        rpcExecutor!!.submit {
            val result = invokeRpc(context, clientToServer.methodName, arguments.value)
            sendReply(clientToServer.replyId, clientToServer.clientAddress, result)
        }
    }
    is Try.Failure -> {
        context = artemisMessage.context(clientToServer.sessionId, emptyList())
        context.invocation.pushToLoggingContext()
        // We failed to deserialise the arguments, route back the error
        log.warn("Inbound RPC failed", arguments.exception)
        sendReply(clientToServer.replyId, clientToServer.clientAddress, arguments)
    }
}
```

lankydan committed:
CORDA-3596 Remove arguments from in-memory checkpoint

When executing a flows first real suspend (from flow code) the arguments
 contained in the `InvocationContext` are removed. This saves holding
 these arguments for the whole lifecyle of a flow.

rick-r3 reviewed (commented):
Looks a lot neater to me.

Have you thought about how you might reset the args or generally avoid persisting them in each subsequent checkpoint ?

lankydan commented:
> Looks a lot neater to me.
> 
> Have you thought about how you might reset the args or generally avoid persisting them in each subsequent checkpoint ?

I have just pushed a commit for this :)

lankydan committed:
CORDA-3596 Increase `cordapp_name` column to 128

lankydan committed:
CORDA-3596 Detekt warning

lankydan committed:
CORDA-3596 Fix `DBCheckpointStorageTest`

lankydan committed:
CORDA-3596 Join metadata by `flow_id`

Due to changes in where metadata is recorded, there is no need for
having `invocation_id` as the metadata table's primary key. The
`flow_id` is now the primary key of the table and is used to join to the
 main checkpoints table.

The `invocation_id` has been removed from the checkpoints table since it
 is not needed for the join anymore.

lankydan committed:
CORDA-3596 `FlowMetadataRecorderTest` -> `FlowMetadataRecordingTest`

lankydan committed:
CORDA-3596 Remove `received_time` from metadata table

lankydan committed:
CORDA-3596 Always set the logging context when handling RPC message

lankydan committed:
CORDA-3596 Remove some warnings

lankydan committed:
CORDA-3596 Remove unused `StartReason` enum

lankydan committed:
CORDA-3596 Remove some warnings

lankydan committed:
CORDA-3596 Simple `DBCheckpointStorageTests` for metadata

lankydan committed:
CORDA-3596 Detekt warning

rick-r3 reviewed (dismissed):
LGTM

lankydan committed:
CORDA-3596 Fix API breakages in `InvocationContext`

lankydan committed:
CORDA-3596 Truncate really long flow names
